### PR TITLE
694 Fix access condition check in SubscriptionController

### LINKED ISSUE #694: Users can still access 'subscribe' link directly for closed groups
## Overview
When a group is closed, a user can still go to the 'subscribe' link for that group and still register to that group. This is a bug and needs fixing.

## Details
This happens on a site which uses `hook_og_user_access_alter` to deny permissions for `subscribe` and `subscribe without approval` for the group in question.

### PR DESCRIPTION
**Pull request by varshith**
#694 

## Technical Details
The issue here is that in `SubscriptionController` there is an access check where the return value of `$this->ogAccess->userAccess()` is checked directly. This will always return `true` even for Forbidden access.
We should use `userAccess()->isAllowed()` or 'userAccess()->isForbidden()' inside the if conditions.
This PR fixes this issue. Tested locally for my case and it works as expected after this fix.

### CODE REVIEW

File: src/Controller/SubscriptionController.php [hunk 1]
```diff
@@ -136,7 +136,8 @@ public function subscribe($entity_type_id, EntityInterface $group, OgMembershipT
       return new RedirectResponse($group->toUrl()->setAbsolute(TRUE)->toString());
     }
 
-    if (!$this->ogAccess->userAccess($group, 'subscribe', $user) && !$this->ogAccess->userAccess($group, 'subscribe without approval', $user)) {
+    if ((($access = $this->ogAccess->userAccess($group, 'subscribe', $user)) && $access->isForbidden()) &&
+      (($access = $this->ogAccess->userAccess($group, 'subscribe without approval', $user)) && $access->isForbidden())) {
```
Comment (MPParsley) [inline src/Controller/SubscriptionController.php]: Making this a bit more readable:
```suggestion
    $subscribe = $this->ogAccess->userAccess($group, 'subscribe', $user);
    $subscribeWithoutApproval = $this->ogAccess->userAccess($group, 'subscribe without approval', $user);
    if ($subscribe->isForbidden() && $subscribeWithoutApproval->isForbidden()) {
```

**Review by MPParsley — CHANGES_REQUESTED** [review]

**Comment by varshith** [comment]
@MPParsley 
done.

File: src/Controller/SubscriptionController.php [hunk 2]
```diff
@@ -136,7 +136,9 @@ public function subscribe($entity_type_id, EntityInterface $group, OgMembershipT
       return new RedirectResponse($group->toUrl()->setAbsolute(TRUE)->toString());
     }
 
-    if (!$this->ogAccess->userAccess($group, 'subscribe', $user) && !$this->ogAccess->userAccess($group, 'subscribe without approval', $user)) {
+    $subscribe = $this->ogAccess->userAccess($group, 'subscribe', $user);
+    $subscribeWithoutApproval = $this->ogAccess->userAccess($group, 'subscribe without approval', $user);
```
Comment (MPParsley) [inline src/Controller/SubscriptionController.php]: We don't really need to pass the user if it's the current user either:
```suggestion
    $subscribe = $this->ogAccess->userAccess($group, 'subscribe');
    $subscribeWithoutApproval = $this->ogAccess->userAccess($group, 'subscribe without approval');
```

File: src/Controller/SubscriptionController.php [hunk 3]
```diff
@@ -136,7 +136,9 @@ public function subscribe($entity_type_id, EntityInterface $group, OgMembershipT
       return new RedirectResponse($group->toUrl()->setAbsolute(TRUE)->toString());
     }
 
-    if (!$this->ogAccess->userAccess($group, 'subscribe', $user) && !$this->ogAccess->userAccess($group, 'subscribe without approval', $user)) {
+    $subscribe = $this->ogAccess->userAccess($group, 'subscribe');
+    $subscribeWithoutApproval = $this->ogAccess->userAccess($group, 'subscribe without approval');
```
Comment (amitaibu) [inline src/Controller/SubscriptionController.php]:
```suggestion
    $subscribe_without_approval = $this->ogAccess->userAccess($group, 'subscribe without approval');
```

**Comment by amitaibu** [comment]
@varshith when you'll have time, it would be great to have a test for this as-well

File: src/Controller/SubscriptionController.php [hunk 4]
```diff
@@ -136,7 +136,9 @@ public function subscribe($entity_type_id, EntityInterface $group, OgMembershipT
       return new RedirectResponse($group->toUrl()->setAbsolute(TRUE)->toString());
     }
 
-    if (!$this->ogAccess->userAccess($group, 'subscribe', $user) && !$this->ogAccess->userAccess($group, 'subscribe without approval', $user)) {
+    $subscribe = $this->ogAccess->userAccess($group, 'subscribe');
+    $subscribe_without_approval = $this->ogAccess->userAccess($group, 'subscribe without approval');
+    if ($subscribe->isForbidden() && $subscribeWithoutApproval->isForbidden()) {
```
Comment (MPParsley) [inline src/Controller/SubscriptionController.php]:
```suggestion
    if ($subscribe->isForbidden() && $subscribe_without_approval->isForbidden()) {
```

**Comment by MPParsley** [comment]
@amitaibu, fixed the snake case.

**Review by MPParsley — APPROVED** [review]

File: src/Controller/SubscriptionController.php (hunk 3 above)
Comment (MPParsley) [reply src/Controller/SubscriptionController.php]: @amitaibu, changed the variable to snake case in the if as well. fyi, you can now also give code suggestions on multiple lines at once by clicking and dragging over the lines you want.

File: src/Controller/SubscriptionController.php [hunk 5]
```diff
@@ -86,7 +86,7 @@ public function subscribe($entity_type_id, EntityInterface $group, OgMembershipT
       throw new AccessDeniedHttpException();
     }
 
-    $user = $this->entityTypeManager()->getStorage('user')->load($this->currentUser()->id());
+    $user = $this->currentUser();
```
Comment (MPParsley) [inline src/Controller/SubscriptionController.php]: Exception: TypeError: Argument 2 passed to Drupal\og\MembershipManager::createMembership() must implement interface Drupal\user\UserInterface, instance of Drupal\Core\Session\AccountProxy given, called in /home/travis/build/Gizra/og/src/Og.php on line 190

Seems like this was the reason the user was loaded as such:

```suggestion
    $user = $this->entityTypeManager()->getStorage('user')->load($this->currentUser()->id());
```

**Comment by pfrenssen** [comment]
It would be better to check for `::isAllowed()` rather than for `::isForbidden()` since this would open up a security issue in case a neutral result is returned. If the access result is neutral then the user shouldn't have access either. Only when either of them returns allowed.

**Review by pfrenssen — CHANGES_REQUESTED** [review]
We need to check for `::isAllowed()` rather than for `::isForbidden()`.

**Comment by pfrenssen** [comment]
The previous implementation is not type safe (it is casting an `AccessResult` object to a boolean) so it should have triggered a test failure when we turned on strict typing. This didn't happen so it appears we have no test coverage for this, and this is a security issue since it allows users to join groups without permission.

It would be a good idea to add a test for this so that we can prevent this security issue regressing in the future. We should also be more careful in the future not to accept feature requests without sufficient coverage.

**Comment by pfrenssen** [comment]
Assigning to write a test.

**Comment by pfrenssen** [comment]
I have pushed the test in a separate PR so that we can check that it fails correctly: #698. This can be merged into this PR once the test results are in.

**Comment by pfrenssen** [comment]
Test is failing correctly, so I merged it in here. Now hopefully the test should turn green.

**Comment by pfrenssen** [comment]
Test is green! I have also changed the check to `::isAllowed()` as I suggested earlier. Unassigning.

This is good to go for me, but I won't merge right now since I added the test.

@varshith thanks a lot for the bug report and the fix!

**Review by pfrenssen — APPROVED** [review]

**Review by MPParsley — APPROVED** [review]

**Pull request merged**